### PR TITLE
Add support for sending and receiving the auth token via a server cookie

### PR DESCRIPTION
**Pull request by theblang**
There are two main reasons to use a server cookie to send and receive the auth token.

1) For sharing auth state across subdomains with the `Domain` attribute. From [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies):

> The Domain attribute specifies which hosts are allowed to receive the cookie. If unspecified, it defaults to the same origin that set the cookie, excluding subdomains. If Domain is specified, then subdomains are always included. Therefore, specifying Domain is less restrictive than omitting it. However, it can be helpful when subdomains need to share information about a user.

Note that you **cannot** access `localStorage` across subdomains.

2) For security reasons. If you only send the auth token in an `HttpOnly` server cookie then JavaScript can never access it, which helps protect against XSS attacks. These three issues are mostly pertaining to the security aspect: https://github.com/lynndylanhurley/devise_token_auth/issues/1281 https://github.com/lynndylanhurley/devise_token_auth/issues/1321 https://github.com/lynndylanhurley/devise_token_auth/issues/1371

This PR introduces a server cookie as an additive opt-in feature to send and receive the auth token, **which will allow for the subdomain sharing mentioned in point one**. By introducing the option to use a server cookie, we lay the groundwork for addressing the security aspect in point two. **I didn't want to make the changes for point two yet** as that will be more impactful, because we'd need to make it possible to **not** expose the auth token in headers. I imagine if and when we do that work we'll probably add an analogous `headers_enabled` config to hinge off of in code to turn off sending and receiving the auth token in headers.

cc @booleanbetrayal @nbrustein @apuntovanini

### CODE REVIEW

File: app/controllers/devise_token_auth/sessions_controller.rb [hunk 1]
```diff
@@ -55,6 +55,16 @@ def destroy
         user.tokens.delete(client)
         user.save!
 
+        if DeviseTokenAuth.cookie_enabled
+          if DeviseTokenAuth.cookie_attributes[:domain]
```
Comment (theblang) [inline app/controllers/devise_token_auth/sessions_controller.rb:59]: I just realized that we might need to be smarter about the `domain` attribute. We might need to determine it by looking at the request's domain (in which case we'd probably want to allow configuring a whitelist, like we do with `redirect_whitelist`). This is because the `Domain` cookie attribute only allows one value, so as the code sits now you wouldn't be able to handle multiple domains being served from the same box.

**Comment by apuntovanini** [comment]
I tested in our application and works like charm. I only have issues logging out (we use a graphql endpoint to make our calls, we should probably switch to classic controllers flow for authentication instead). But sharing cookies with our subdomains is crucial to us, well done!

File: app/controllers/devise_token_auth/sessions_controller.rb (hunk 1 above)
Comment (apuntovanini) [reply app/controllers/devise_token_auth/sessions_controller.rb:59]: Make sense, remember to use allowlist instead of whitelist

Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb:59]: @apuntovanini Another option is to just support passing to the `cookie_attributes[:domain]` config either a string, which will work for most people, or a proc, which could accept the request and do whatever logic (in our case, return the domain of the request so long as it matches the allow list).

**Comment by theblang** [comment]
> I tested in our application and works like charm. I only have issues logging out (we use a graphql endpoint to make our calls, we should probably switch to classic controllers flow for authentication instead). But sharing cookies with our subdomains is crucial to us, well done!

Yay, thanks so much! I'm so glad this will help y'all as well!

File: app/controllers/devise_token_auth/sessions_controller.rb (hunk 1 above)
Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb]: [Here](https://github.com/lynndylanhurley/devise_token_auth/pull/1453/commits/7592336caa29f8a1c4d9173f409993becba6e1ef) are the changes for that second strategy. And my config looks like:

```
    config.cookie_attributes = {
        domain: lambda { |request| request.domain if valid_cookie_domains.include?(request.domain) },
        secure: Rails.env.production?,
        httponly: true,
        same_site: 'Strict',
        expires: token_lifespan.from_now,
        encrypt: true
    }
```

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb [hunk 2]
```diff
@@ -113,6 +126,18 @@ def update_auth_header
...
+    # Most people will just set a string for the domain attribute config. But if you need
+    # more flexibility, such as to dynamically choose the domain when serving multiple domains
+    # from a single server, you can set a Proc that will be passed the request.
+    config_value = DeviseTokenAuth.cookie_attributes[:domain]
+    if config_value.is_a?(String)
+      config_value
+    elsif config_value.is_a?(Proc)
+      config_value.call(request)
+    end
+  end
```
Comment (theblang) [inline app/controllers/devise_token_auth/concerns/set_user_by_token.rb:139]: See [this discussion](https://github.com/lynndylanhurley/devise_token_auth/pull/1453#discussion_r552158961)

File: .gitignore [hunk 3]
```diff
@@ -9,6 +9,7 @@ test/dummy/.sass-cache
 test/dummy/config/application.yml
 coverage
 .idea
+.byebug_history
```
Comment (theblang) [inline .gitignore:12]: Drive-by that popped up while debugging a test

File: app/controllers/devise_token_auth/sessions_controller.rb [hunk 4]
```diff
@@ -55,6 +55,17 @@ def destroy
         user.tokens.delete(client)
         user.save!
 
+        if DeviseTokenAuth.cookie_enabled
+          cookie_domain = get_cookie_domain
+          if cookie_domain.present?
+            # If a cookie is set with a domain specified then it must be deleted with that domain specified
+            # See https://stackoverflow.com/a/6244724/1747491
+            cookies.delete(DeviseTokenAuth.cookie_name, domain: cookie_domain)
```
Comment (MaicolBen) [inline app/controllers/devise_token_auth/sessions_controller.rb:63]: This line is unreached by specs but it could be ok

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb [hunk 5]
```diff
@@ -123,11 +148,30 @@ def refresh_headers
...
     end # end lock
   end
 
+  def set_cookie(auth_header)
+    cookie_attributes = DeviseTokenAuth.cookie_attributes
+    cookie_attributes.merge!(value: auth_header.to_json)
+
+    cookie_domain = get_cookie_domain
+    if cookie_domain.present?
+      cookie_attributes.merge!(domain: cookie_domain)
```
Comment (MaicolBen) [inline app/controllers/devise_token_auth/concerns/set_user_by_token.rb:169]: This line is unreached by specs but it could be ok

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb [hunk 6]
```diff
@@ -113,6 +126,18 @@ def update_auth_header
     end
   end
 
+  def get_cookie_domain
+    # Most people will just set a string for the domain attribute config. But if you need
+    # more flexibility, such as to dynamically choose the domain when serving multiple domains
+    # from a single server, you can set a Proc that will be passed the request.
+    config_value = DeviseTokenAuth.cookie_attributes[:domain]
+    if config_value.is_a?(String)
```
Comment (MaicolBen) [inline app/controllers/devise_token_auth/concerns/set_user_by_token.rb:134]: These cases are unreached in specs but that could be ok

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb [hunk 7]
```diff
@@ -35,11 +35,20 @@ def set_user_by_token(mapping = nil)
     access_token_name = DeviseTokenAuth.headers_names[:'access-token']
     client_name = DeviseTokenAuth.headers_names[:'client']
 
+    # gets values from cookie if configured and present
```
Comment (MaicolBen) [inline app/controllers/devise_token_auth/concerns/set_user_by_token.rb:38]: can all this new cookie code be in a new module like `set_user_by_cookie` or something like that, so we don't pollute more this big file, you can include the concern/module in this file

**Review by MaicolBen — COMMENTED** [review]
I don't like using cookies in APIs but your reasoning of why is a good alternative sounds good

**Comment by apuntovanini** [comment]
> I don't like using cookies in APIs but your reasoning of why is a good alternative sounds good

Just to understand, how would how tackle cross-subdomain authentication with api without a cookie? Thanks!

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb (hunk 5 above)
Comment (theblang) [reply app/controllers/devise_token_auth/concerns/set_user_by_token.rb]: This conditional will be moot now, see [this recent commit](https://github.com/lynndylanhurley/devise_token_auth/pull/1453/commits/536340b3e1861fef0e595158b3588b747d765898).

File: app/controllers/devise_token_auth/sessions_controller.rb (hunk 4 above)
Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb]: This conditional will be moot now, see [this recent commit](https://github.com/lynndylanhurley/devise_token_auth/pull/1453/commits/536340b3e1861fef0e595158b3588b747d765898).

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb (hunk 7 above)
Comment (theblang) [reply app/controllers/devise_token_auth/concerns/set_user_by_token.rb:38]: > I don't like using cookies in APIs but your reasoning of why is a good alternative sounds good

Yeah, we weren't too keen on it either. But it's the only way to share state across subdomains. And the general consensus seems to be that it's more secure for auth (once you're using **only** the HttpOnly cookie, which we'll hopefully allow in later work).

**Comment by theblang** [comment]
> I don't like using cookies in APIs but your reasoning of why is a good alternative sounds good

Yeah, we weren't too keen on it either. But it's the only way to share state across subdomains. And the general consensus seems to be that it's more secure for auth (once you're using **only** the HttpOnly cookie, which we'll hopefully allow in later work).

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb (hunk 7 above)
Comment (theblang) [reply app/controllers/devise_token_auth/concerns/set_user_by_token.rb:38]: I can, my only push back would be that hopefully in the future we'll allow turning off headers (for the security reasons mentioned in point two of the description). In that case, it might feel a bit disjointed to have the header logic in `set_user_by_token` while the cookie logic is in `set_user_by_cookie`, when really both logics deal with handling the token. In that future world, would we move the header logic to another concern called `set_user_by_headers`? If so, wouldn't `set_user_by_token` just become a skeleton file?

File: app/controllers/devise_token_auth/sessions_controller.rb (hunk 1 above)
Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb]: 🤦  I just saw this in the [in the Rails docs](https://api.rubyonrails.org/classes/ActionDispatch/Cookies.html):

> To support multiple domains, provide an array, and the first domain matching request.host will be used. Make sure to specify the :domain option with :all or Array again when deleting cookies.

So I can totally gut this Proc logic now, and instead just pass along either a string or Array from the config to the cookie call.

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb (hunk 7 above)
Comment (theblang) [reply app/controllers/devise_token_auth/concerns/set_user_by_token.rb:38]: On this note, see [my recent realization](https://github.com/lynndylanhurley/devise_token_auth/pull/1453#discussion_r553446626). That will clean up the amount of noise in this file considerably, commit incoming.

File: app/controllers/devise_token_auth/sessions_controller.rb (hunk 1 above)
Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb]: Actually, using an array doesn't quite work for us since the `request.host` that it matches against includes the subdomain. But I think the `:all` option will:

> If you use a schema like www.example.com and want to share session with user.example.com set :domain to :all

Comment (theblang) [reply app/controllers/devise_token_auth/sessions_controller.rb]: See [this commit](https://github.com/lynndylanhurley/devise_token_auth/pull/1453/commits/66f8fa98cfbefbeaa5f098f3efa482744f9e668d). So much cleaner! I wish I had realized that earlier, lol.

File: app/controllers/devise_token_auth/concerns/set_user_by_token.rb (hunk 6 above)
Comment (theblang) [reply app/controllers/devise_token_auth/concerns/set_user_by_token.rb]: Moot now that I've gutted the unnecessary Proc logic.

**Comment by theblang** [comment]
@MaicolBen @booleanbetrayal Could one of you retry the failed Travis job? It seems that just one of the jobs failed with:

```
Fetching nokogiri 1.10.10
Bundler::GemspecError: Could not read gem at
/home/travis/build/lynndylanhurley/devise_token_auth/gemfiles/vendor/bundle/ruby/2.3.0/cache/nokogiri-1.10.10.gem.
It may be corrupted.
An error occurred while installing nokogiri (1.10.10), and Bundler
cannot continue.
Make sure that `gem install nokogiri -v '1.10.10' --source
'https://rubygems.org/'` succeeds before bundling.
In rails_5_1.gemfile:
  devise_token_auth was resolved to 1.1.4, which depends on
    rails was resolved to 5.2.4.4, which depends on
      actioncable was resolved to 5.2.4.4, which depends on
        actionpack was resolved to 5.2.4.4, which depends on
          actionview was resolved to 5.2.4.4, which depends on
            rails-dom-testing was resolved to 2.0.3, which depends on
              nokogiri
```

**Comment by booleanbetrayal** [comment]
@theblang - running this now produces a new error: 

`You are using an old OmniAuth version, please ensure you have 1.0.0.pr2 version or later installed.`

It looks like Devise has [hard-coded the omniauth dependency](https://github.com/heartcombo/devise/issues/5326) in a way that it's breaking it's 2.0.0 release. It would be preferable if they fixed this on their end, but we might want to constrain the Gem version in the project if they are slow with a release. (cc: @MaicolBen )

**Comment by MaicolBen** [comment]
I'm going to wait for https://github.com/heartcombo/devise/issues/5326 for a few days

**Comment by theblang** [comment]
[Looks like](https://github.com/heartcombo/devise/pull/5327#issuecomment-769811843) we're close!

**Comment by MaicolBen** [comment]
Can you rebase/update with last master?
